### ISSUE TITLE
Time and time zones

### ISSUE THREAD

**Issue by malx122** [Starcounter/Starcounter#3674]
What is the recommended way to deal with time and time zones so its correctly handled by ours and third party apps together? Should all SC servers and data be in UTC and all conversion on the clients? Or something else? Quite important if you should start map data together.

**Comment by miyconst** [comment]
Unfortunately we don't yet have a recommended way of doing this in Starcounter, but I can share some experience I have.

There are multiple types of date time fields:

- Date without time and time zone. For example birth date of a person. It should be displayed the same regardless of Client's time zone.
- Date time value. For example date time displayed on a photo. It should be displayed the same for all users.
- Point of time. For example an event info - start, end date time. It should be displayed in the User's time zone.

That being said, we cannot blindly convert all dates to UTC and back to User's timezone. The `point of time` values should be specially treated and converted to/from UTC.

Some points on the implementation:

- It is incorrect to rely on Client's time zone. The time zone settings should be saved per each user.
- It should not matter what time zone the server is, the database values should always be the same.
- It is nice to mark all of the time zone related properties, for example add `Utc` in the name. So `StartDate` became `StartDateUtc`, while `StartDate` will be a getter/setter to convert value from/to user's time to/from UTC.

**Comment by malx122** [comment]
Yes, I know my preferred way of doing this. It is that all datetime in the db should be in UTC and everything else handled by the apps (Probably at the client since the client is the best one deciding timezone). But if some applications (3:rd party) store in server time and our apps in UTC and even one more in something else, there will be impossible to map together. Therefore there MUST be a standard set if SC apps should ever work together.

I actually only see two solutions. Either you store the timezone information together with the value in the db (and then you need to send it with every automapped json property) or decide that everything always stored in UTC and it is UTC values that is auto mapped.

This also affects your version 2.4 and the log feature if I'm not mistaken.

**Comment by miyconst** [comment]
>Either you store the timezone information together with the value in the db (and then you need to send it with every automapped json property) or decide that everything always stored in UTC and it is UTC values that is auto mapped.

I don't see how that can be achieved.

- How to know what is the time zone of a particular value being saved?
- How to know that a value coming from an app is not yet converted into UTC?
- How to know which values to convert and which not?

In the future the database mappers will be moved outside of the apps. It will be up to the mappers to decide which value to convert from which time zone to which time zone when merging data between two or more apps.

If you need a timezone for particular date, then you can either save it as a separate column, or create a `struct` to save date time and time zone values.

This is a more global problem, not only time zone. For example `Length`, it can be measured in `km` or `mil`. There is no way to define a global length measure unit and automatically convert all of the length values when saved to database. If we want two apps to work, one of which uses `km` and another one `mil`, we have to specify correct length converter inside the database mapper.

**Comment by malx122** [comment]
You can't compare time zones with units, that is two totally different things. One is scaled, the other one is moved. First one its reasonable to be able to take by looking at the tables, the second one not. Do you expect the mapping person to be in contact with all developers of third party apps?

**Comment by miyconst** [comment]
@malx122 I am not comparing time zone to measure units, just mentioned it as another issue to be solved by mapping.

> Do you expect the mapping person to be in contact with all developers of third party apps?

No, but each app developer should provide a default map with full description for his app. If an app lacks the default map, or the default map is not good enough, such app won't be accepted to the Warehouse.

**Comment by dan31** [comment]
Related: https://github.com/Starcounter/Starcounter/issues/3706
